<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />


<meta name="author" content="Jason Willwerscheid" />


<title>Variance Structures</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>
<script>
$(document).ready(function () {
  window.initializeCodeFolding("hide" === "show");
});
</script>



<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}


.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
  padding-left: 25px;
  text-indent: 0;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">scFLASH</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/willwerscheid/scFLASH">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>

<div class="fluid-row" id="header">

<div class="btn-group pull-right">
<button type="button" class="btn btn-default btn-xs dropdown-toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu" style="min-width: 50px;">
<li><a id="rmd-show-all-code" href="#">Show All Code</a></li>
<li><a id="rmd-hide-all-code" href="#">Hide All Code</a></li>
</ul>
</div>



<h1 class="title toc-ignore">Variance Structures</h1>
<h4 class="author">Jason Willwerscheid</h4>
<h4 class="date">8/12/2019</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span> workflowr <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#report"> Report <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2019-09-04
</p>
<p>
<strong>Checks:</strong> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 6 <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>scFLASH/</code> <span class="glyphicon glyphicon-question-sign" aria-hidden="true" title="This is the local directory in which the code in this file was executed."> </span>
</p>
<p>
This reproducible <a href="http://rmarkdown.rstudio.com">R Markdown</a> analysis was created with <a
  href="https://github.com/jdblischak/workflowr">workflowr</a> (version 1.2.0). The <em>Report</em> tab describes the reproducibility checks that were applied when the results were created. The <em>Past versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="report" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date </a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate" class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git repository, you know the exact version of the code that produced these results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the global environment can affect the analysis in your R Markdown file in unknown ways. For reproduciblity it’s best to always run the code in an empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20181103code"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Seed:</strong> <code>set.seed(20181103)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20181103code" class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20181103)</code> was run prior to running the code in the R Markdown file. Setting a seed ensures that any results that rely on randomness, e.g. subsampling or permutations, are reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Session information:</strong> recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be confident that you successfully produced the results during this run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomwillwerscheidscFLASHtreeb8fe277ae8af35278603ab0c883df50a3213a9a5targetblankb8fe277a"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Repository version:</strong> <a href="https://github.com/willwerscheid/scFLASH/tree/b8fe277ae8af35278603ab0c883df50a3213a9a5" target="_blank">b8fe277</a> </a>
</p>
</div>
<div id="strongRepositoryversionstrongahrefhttpsgithubcomwillwerscheidscFLASHtreeb8fe277ae8af35278603ab0c883df50a3213a9a5targetblankb8fe277a" class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility. The version displayed above was the version of the Git repository at the time these results were generated. <br><br> Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .DS_Store
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/
    Ignored:    code/initialization/
    Ignored:    data/droplet.rds
    Ignored:    output/backfit/
    Ignored:    output/prior_type/
    Ignored:    output/size_factors/
    Ignored:    output/var_type/

Untracked files:
    Untracked:  analysis/NBapprox.Rmd
    Untracked:  analysis/deleted.Rmd
    Untracked:  analysis/trachea4.Rmd
    Untracked:  code/missing_data.R
    Untracked:  code/pseudocount/
    Untracked:  code/pseudocounts.R
    Untracked:  code/trachea4.R
    Untracked:  data/Ensembl2Reactome.txt
    Untracked:  data/hard_bimodal1.txt
    Untracked:  data/hard_bimodal2.txt
    Untracked:  data/hard_bimodal3.txt
    Untracked:  data/mus_pathways.rds
    Untracked:  docs/figure/pseudocount2.Rmd/
    Untracked:  output/pseudocount/

Unstaged changes:
    Modified:   analysis/index.Rmd
    Modified:   analysis/pseudocount.Rmd
    Modified:   code/sc_comparisons.R
    Modified:   code/utils.R

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the R Markdown and HTML files. If you’ve configured a remote Git repository (see <code>?wflow_git_remote</code>), click on the hyperlinks in the table below to view them.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/willwerscheid/scFLASH/blob/b8fe277ae8af35278603ab0c883df50a3213a9a5/analysis/var_type.Rmd" target="_blank">b8fe277</a>
</td>
<td>
Jason Willwerscheid
</td>
<td>
2019-09-04
</td>
<td>
wflow_publish(“analysis/var_type.Rmd”)
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/willwerscheid/scFLASH/9c8fdf9f3df59dd28461e450d639a009bdc77425/docs/var_type.html" target="_blank">9c8fdf9</a>
</td>
<td>
Jason Willwerscheid
</td>
<td>
2019-09-03
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/willwerscheid/scFLASH/blob/723ee8818b2172ecce7829d77d0977acc6815740/analysis/var_type.Rmd" target="_blank">723ee88</a>
</td>
<td>
Jason Willwerscheid
</td>
<td>
2019-09-03
</td>
<td>
wflow_publish(“analysis/var_type.Rmd”)
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/willwerscheid/scFLASH/851baf81070c51d052c48571d7c2cea5be742bda/docs/var_type.html" target="_blank">851baf8</a>
</td>
<td>
Jason Willwerscheid
</td>
<td>
2019-09-01
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/willwerscheid/scFLASH/blob/fc78232542efa2327891106309580aa8f276b6c2/analysis/var_type.Rmd" target="_blank">fc78232</a>
</td>
<td>
Jason Willwerscheid
</td>
<td>
2019-09-01
</td>
<td>
wflow_publish(c(“analysis/backfit.Rmd”, “analysis/var_type.Rmd”))
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/willwerscheid/scFLASH/e2d99bb92ebb6b20e8700aefcb5d2bd837c2e290/docs/var_type.html" target="_blank">e2d99bb</a>
</td>
<td>
Jason Willwerscheid
</td>
<td>
2019-08-31
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/willwerscheid/scFLASH/blob/f608bb0637b4a4d6972f210ca9454dffc8a94f12/analysis/var_type.Rmd" target="_blank">f608bb0</a>
</td>
<td>
Jason Willwerscheid
</td>
<td>
2019-08-31
</td>
<td>
wflow_publish(“analysis/var_type.Rmd”)
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/willwerscheid/scFLASH/26d29a37b931efee69f83e3c05aa72a401c96948/docs/var_type.html" target="_blank">26d29a3</a>
</td>
<td>
Jason Willwerscheid
</td>
<td>
2019-08-16
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/willwerscheid/scFLASH/blob/bd61d1a88f7a97dea755963c052ae4105bc05e79/analysis/var_type.Rmd" target="_blank">bd61d1a</a>
</td>
<td>
Jason Willwerscheid
</td>
<td>
2019-08-16
</td>
<td>
wflow_publish(“analysis/var_type.Rmd”)
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="introduction" class="section level2">
<h2>Introduction</h2>
<p>Recall that EBMF fits the model <span class="math display">\[ Y_{ij} = \sum_k L_{ik}F_{jk} + E_{ij}, \]</span> where <span class="math display">\[ E_{ij} \sim N(0, \sigma^2_{ij}) \]</span> By assuming that the matrix <span class="math inline">\(\Sigma\)</span> has a given structure, one can estimate the <span class="math inline">\(\sigma_{ij}^2\)</span>s via maximum likelihood. In <code>flashier</code>, one can assume:</p>
<ul>
<li>A constant variance structure: <span class="math inline">\(\sigma_{ij}^2 = \sigma^2\)</span>.</li>
<li>A gene-wise variance structure: <span class="math inline">\(\sigma_{ij}^2 = \sigma_i^2\)</span>.</li>
<li>A cell-wise variance structure: <span class="math inline">\(\sigma_{ij}^2 = \sigma_j^2\)</span>.</li>
<li>A “Kronecker” (rank-one) variance structure: <span class="math inline">\(\sigma_{ij}^2 = \sigma_{(1)i}^2 \sigma_{(2)j}^2\)</span>.</li>
</ul>
<p>The question, then, is how much the “errors” in the transformed data matrix <span class="math inline">\(Y\)</span> vary from gene to gene and from cell to cell, and whether accounting for this heteroskedasticity makes any difference to the quality of the factors obtained via <code>flashier</code>.</p>
<p>In this analysis, I’ll argue that accounting for differences in variability among both genes and cells is crucial, but I’ll advocate for a quick approximation to a Kronecker variance structure (which I’ll describe shortly) rather than a full Kronecker fit.</p>
<p>All fits were produced by adding 20 “greedy” factors to the <a href="droplet.html">Montoro et al. droplet dataset</a>. The code can be viewed <a href="https://github.com/willwerscheid/scFLASH/blob/master/code/var_type/vartype_fits.R">here</a>.</p>
<pre class="r"><code>source(&quot;./code/utils.R&quot;)
droplet &lt;- readRDS(&quot;./data/droplet.rds&quot;)
droplet &lt;- preprocess.droplet(droplet)
res &lt;- readRDS(&quot;./output/var_type/vartype_fits.rds&quot;)</code></pre>
</div>
<div id="gene-wise-variance-and-sparsity" class="section level2">
<h2>Gene-wise variance and sparsity</h2>
<p>I calculate the “sparsity” of a gene as the proportion of cells that have a nonzero number of transcripts for that gene. (I prefer this measure to, say, mean expression because it doesn’t depend on how size factors are calculated.)</p>
<p>There is a very clear relationship between sparsity and gene-wise precision: in general, sparser genes yield more precise estimates. This is probably not surprising. If a gene is very sparse, then most estimates will be close to zero with high precision.</p>
<p>The reversal of this trend for the least sparse genes is more surprising. I think that this is at least partially an effect of library-size normalization: since library size is strongly correlated with the expression levels of the most highly expressed genes, the variability of the least sparse genes can be (artificially) reduced by scaling. I will investigate the effects of using size factors to normalize counts in a <a href="size_factors.html">subsequent analysis</a>.</p>
<pre class="r"><code>gene.df &lt;- data.frame(sparsity = droplet$gene.sparsity,
                      genewise.sd = res$bygene$fl$residuals.sd,
                      kronecker.sd = res$kron$fl$residuals.sd[[1]],
                      approx.sd = res$scale.both$gene.prescaling.factors)
ggplot(gene.df, aes(x = sparsity, y = genewise.sd)) +
  geom_point(alpha = 0.2) +
  labs(y = &quot;estimated sd&quot;)</code></pre>
<p><img src="figure/var_type.Rmd/gw-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-gw-1">
Past versions of gw-1.png
</button>
</p>
<div id="fig-gw-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/willwerscheid/scFLASH/blob/e2d99bb92ebb6b20e8700aefcb5d2bd837c2e290/docs/figure/var_type.Rmd/gw-1.png" target="_blank">e2d99bb</a>
</td>
<td>
Jason Willwerscheid
</td>
<td>
2019-08-31
</td>
</tr>
<tr>
<td>
<a href="https://github.com/willwerscheid/scFLASH/blob/26d29a37b931efee69f83e3c05aa72a401c96948/docs/figure/var_type.Rmd/gw-1.png" target="_blank">26d29a3</a>
</td>
<td>
Jason Willwerscheid
</td>
<td>
2019-08-16
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="cell-wise-variance-and-library-size" class="section level2">
<h2>Cell-wise variance and library size</h2>
<p>Again the relationship is clear: a larger library size yields higher precision. This is to be expected, since a larger library size essentially means that we have more and better information about a cell.</p>
<pre class="r"><code>cell.df &lt;- data.frame(library.size = droplet$lib.size,
                      cellwise.sd = res$bycell$fl$residuals.sd,
                      kronecker.sd = res$kron$fl$residuals.sd[[2]],
                      approx.sd = res$scale.both$cell.prescaling.factors)
ggplot(cell.df, aes(x = library.size, y = cellwise.sd)) +
  geom_point(alpha = 0.2) +
  labs(x = &quot;library size&quot;, y = &quot;estimated sd&quot;)</code></pre>
<p><img src="figure/var_type.Rmd/cw-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-cw-1">
Past versions of cw-1.png
</button>
</p>
<div id="fig-cw-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/willwerscheid/scFLASH/blob/e2d99bb92ebb6b20e8700aefcb5d2bd837c2e290/docs/figure/var_type.Rmd/cw-1.png" target="_blank">e2d99bb</a>
</td>
<td>
Jason Willwerscheid
</td>
<td>
2019-08-31
</td>
</tr>
<tr>
<td>
<a href="https://github.com/willwerscheid/scFLASH/blob/26d29a37b931efee69f83e3c05aa72a401c96948/docs/figure/var_type.Rmd/cw-1.png" target="_blank">26d29a3</a>
</td>
<td>
Jason Willwerscheid
</td>
<td>
2019-08-16
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="kronecker-variance-and-pre-scaling" class="section level2">
<h2>Kronecker variance and pre-scaling</h2>
<p>A Kronecker variance structure allows us to account for both gene-wise and cell-wise variance, but it has the disadvantage of being much slower to fit.</p>
<p>To get most of the benefits of the Kronecker variance structure without sacrificing too much in performance, notice that the Kronecker model <span class="math display">\[ Y_{ij} = \sum_k L_{ik}F_{jk} + E_{ij},\ 
E_{ij} \sim N(0, \sigma_i^{2(1)} \sigma_j^{2(2)}) \]</span> is nearly equivalent to the model <span class="math display">\[ Y_{ij} / \sigma_j^{(2)} = \sum_k L_{ik}\frac{F_{jk}}{\sigma_j^{(2)}} + E_{ij},\
E_{ij} \sim N(0, \sigma_i^{2(1)}), \]</span> so that, if one has cell-wise variance estimates, one can simply scale the columns in advance and then fit a gene-wise variance structure.</p>
<p>The difference is that the meaning of the prior on <span class="math inline">\(F\)</span> (which I have ignored until now) changes. Instead of fitting empirical Bayes priors <span class="math display">\[ F_{jk} \sim g_k^{(f)}, \]</span> the pre-scaling approach fits the “p-value prior” <span class="math display">\[ \frac{F_{jk}}{\sigma_j^{(2)}} \sim g_k^{(f)}. \]</span> Similarly, by scaling the rows in advance, it is possible to fit <span class="math display">\[ \frac{L_{ik}}{\sigma_i^{(1)}} \sim g_k^{(\ell)}. \]</span> These observations suggest three possible ways of proceeding: scaling the columns in advance and fitting gene-wise variance estimates, scaling the rows and fitting cell-wise estimates, and scaling both and fitting gene-wise estimates. (One could also scale both and fit cell-wise estimates, but I’m convinced that getting variance estimates right for genes is more important than for cells: see <a href="var_type.html#comparisons:_elbo">below</a>.) All approaches have the important advantage of being able to account for heteroskedasticity among both genes and cells, but they require nearly as little time to fit as the simpler cell- and gene-wise approaches (some additional time is needed to obtain the initial scaling factors):</p>
<pre class="r"><code>t.df &lt;- data.frame(elapsed.time = sapply(res, `[[`, &quot;elapsed.time&quot;))
t.df$fit &lt;- names(res)
the.limits &lt;- c(&quot;const&quot;, &quot;bygene&quot;, &quot;bycell&quot;, 
                &quot;scale.genes&quot;, &quot;scale.cells&quot;, &quot;scale.both&quot;, &quot;kron&quot;)
ggplot(t.df, aes(x = fit, y = elapsed.time)) + 
  geom_bar(stat = &quot;identity&quot;) +
  scale_x_discrete(limits = the.limits) +
  scale_y_log10() +
  labs(x = NULL, y = &quot;elapsed time (s)&quot;)</code></pre>
<p><img src="figure/var_type.Rmd/timing-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-timing-1">
Past versions of timing-1.png
</button>
</p>
<div id="fig-timing-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/willwerscheid/scFLASH/blob/9c8fdf9f3df59dd28461e450d639a009bdc77425/docs/figure/var_type.Rmd/timing-1.png" target="_blank">9c8fdf9</a>
</td>
<td>
Jason Willwerscheid
</td>
<td>
2019-09-03
</td>
</tr>
<tr>
<td>
<a href="https://github.com/willwerscheid/scFLASH/blob/e2d99bb92ebb6b20e8700aefcb5d2bd837c2e290/docs/figure/var_type.Rmd/timing-1.png" target="_blank">e2d99bb</a>
</td>
<td>
Jason Willwerscheid
</td>
<td>
2019-08-31
</td>
</tr>
<tr>
<td>
<a href="https://github.com/willwerscheid/scFLASH/blob/26d29a37b931efee69f83e3c05aa72a401c96948/docs/figure/var_type.Rmd/timing-1.png" target="_blank">26d29a3</a>
</td>
<td>
Jason Willwerscheid
</td>
<td>
2019-08-16
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>In this analysis, I fit a rank-one Kronecker model to get initial gene- and cell-wise variance estimates. Indeed, these estimates track those obtained using a full Kronecker fit fairly closely (note that the estimates can all be scaled by a constant without changing the model, so I’m looking for a linear relationship here rather than the relationship <span class="math inline">\(y = x\)</span>).</p>
<pre class="r"><code>ggplot(gene.df, aes(x = kronecker.sd, y = approx.sd)) +
  geom_point(alpha = 0.2) +
  labs(x = &quot;gene sd (full Kronecker)&quot;, y = &quot;gene sd (rank-one Kronecker)&quot;)</code></pre>
<p><img src="figure/var_type.Rmd/kron.sds-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-kron.sds-1">
Past versions of kron.sds-1.png
</button>
</p>
<div id="fig-kron.sds-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/willwerscheid/scFLASH/blob/e2d99bb92ebb6b20e8700aefcb5d2bd837c2e290/docs/figure/var_type.Rmd/kron.sds-1.png" target="_blank">e2d99bb</a>
</td>
<td>
Jason Willwerscheid
</td>
<td>
2019-08-31
</td>
</tr>
<tr>
<td>
<a href="https://github.com/willwerscheid/scFLASH/blob/26d29a37b931efee69f83e3c05aa72a401c96948/docs/figure/var_type.Rmd/kron.sds-1.png" target="_blank">26d29a3</a>
</td>
<td>
Jason Willwerscheid
</td>
<td>
2019-08-16
</td>
</tr>
</tbody>
</table>
</div>
</div>
<pre class="r"><code>ggplot(cell.df, aes(x = kronecker.sd, y = approx.sd)) +
  geom_point(alpha = 0.2) +
  labs(x = &quot;cell sd (full Kronecker)&quot;, y = &quot;cell sd (rank-one Kronecker)&quot;)</code></pre>
<p><img src="figure/var_type.Rmd/kron.sds-2.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-kron.sds-2">
Past versions of kron.sds-2.png
</button>
</p>
<div id="fig-kron.sds-2" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/willwerscheid/scFLASH/blob/e2d99bb92ebb6b20e8700aefcb5d2bd837c2e290/docs/figure/var_type.Rmd/kron.sds-2.png" target="_blank">e2d99bb</a>
</td>
<td>
Jason Willwerscheid
</td>
<td>
2019-08-31
</td>
</tr>
<tr>
<td>
<a href="https://github.com/willwerscheid/scFLASH/blob/26d29a37b931efee69f83e3c05aa72a401c96948/docs/figure/var_type.Rmd/kron.sds-2.png" target="_blank">26d29a3</a>
</td>
<td>
Jason Willwerscheid
</td>
<td>
2019-08-16
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="results-elbo" class="section level2">
<h2>Results: ELBO</h2>
<p>A comparison of ELBOs suggests that accounting for heteroskedasticity among genes is more important than accounting for differences in variability among cells.</p>
<pre class="r"><code>n.genes &lt;- nrow(droplet$data)
n.cells &lt;- ncol(droplet$data)
elbo &lt;- sapply(res, function(x) x$fl$elbo + x$elbo.adj)
addl.par &lt;- c(n.cells - 1, n.genes - 1, n.genes, n.genes, n.cells, n.cells)
nested.df &lt;- data.frame(comparison = c(&quot;constant -&gt; cell-wise&quot;,
                                       &quot;constant -&gt; gene-wise&quot;,
                                       &quot;cell-wise -&gt; kronecker&quot;,
                                       &quot;cell-wise -&gt; pre-scale genes&quot;,
                                       &quot;gene-wise -&gt; kronecker&quot;,
                                       &quot;gene-wise -&gt; pre-scale cells&quot;),
                        addl.par = addl.par,
                        elbo.per.par = (round(c(elbo[&quot;bycell&quot;] - elbo[&quot;const&quot;],
                                                elbo[&quot;bygene&quot;] - elbo[&quot;const&quot;],
                                                elbo[&quot;kron&quot;] - elbo[&quot;bycell&quot;],
                                                elbo[&quot;scale.genes&quot;] - elbo[&quot;bycell&quot;],
                                                elbo[&quot;kron&quot;] - elbo[&quot;bygene&quot;],
                                                elbo[&quot;scale.cells&quot;] - elbo[&quot;bygene&quot;])
                                              / addl.par)))
col.names =  c(&quot;Nested models&quot;, 
               &quot;Number of additional parameters&quot;, 
               &quot;Improvement in ELBO per parameter&quot;)
knitr::kable(nested.df[1:2, ], col.names = col.names)</code></pre>
<table>
<thead>
<tr class="header">
<th align="left">Nested models</th>
<th align="right">Number of additional parameters</th>
<th align="right">Improvement in ELBO per parameter</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">constant -&gt; cell-wise</td>
<td align="right">7190</td>
<td align="right">893</td>
</tr>
<tr class="even">
<td align="left">constant -&gt; gene-wise</td>
<td align="right">14480</td>
<td align="right">3644</td>
</tr>
</tbody>
</table>
<p>A comparison of the Kronecker fit and pre-scaling approaches (being careful to adjust the ELBO of the latter to account for the scaling of the data) confirms that the quality of the two is similar.</p>
<pre class="r"><code>knitr::kable(nested.df[3:6, ], col.names = col.names, row.names = FALSE)</code></pre>
<table>
<thead>
<tr class="header">
<th align="left">Nested models</th>
<th align="right">Number of additional parameters</th>
<th align="right">Improvement in ELBO per parameter</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">cell-wise -&gt; kronecker</td>
<td align="right">14481</td>
<td align="right">3814</td>
</tr>
<tr class="even">
<td align="left">cell-wise -&gt; pre-scale genes</td>
<td align="right">14481</td>
<td align="right">3789</td>
</tr>
<tr class="odd">
<td align="left">gene-wise -&gt; kronecker</td>
<td align="right">7191</td>
<td align="right">1236</td>
</tr>
<tr class="even">
<td align="left">gene-wise -&gt; pre-scale cells</td>
<td align="right">7191</td>
<td align="right">1227</td>
</tr>
</tbody>
</table>
<p>In fact, pre-scaling both genes and cells yields the best ELBO among all fits.</p>
<pre class="r"><code>elbo.df &lt;- data.frame(fit = names(elbo), elbo = elbo)
ggplot(subset(elbo.df, elbo &gt; 40000000), aes(x = fit, y = elbo)) + 
  geom_point() +
  scale_x_discrete(limits = c(&quot;scale.genes&quot;, &quot;scale.cells&quot;, &quot;scale.both&quot;, &quot;kron&quot;)) +
  labs(x = NULL, y = &quot;ELBO&quot;)</code></pre>
<p><img src="figure/var_type.Rmd/kron.compare-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-kron.compare-1">
Past versions of kron.compare-1.png
</button>
</p>
<div id="fig-kron.compare-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/willwerscheid/scFLASH/blob/9c8fdf9f3df59dd28461e450d639a009bdc77425/docs/figure/var_type.Rmd/kron.compare-1.png" target="_blank">9c8fdf9</a>
</td>
<td>
Jason Willwerscheid
</td>
<td>
2019-09-03
</td>
</tr>
<tr>
<td>
<a href="https://github.com/willwerscheid/scFLASH/blob/e2d99bb92ebb6b20e8700aefcb5d2bd837c2e290/docs/figure/var_type.Rmd/kron.compare-1.png" target="_blank">e2d99bb</a>
</td>
<td>
Jason Willwerscheid
</td>
<td>
2019-08-31
</td>
</tr>
<tr>
<td>
<a href="https://github.com/willwerscheid/scFLASH/blob/26d29a37b931efee69f83e3c05aa72a401c96948/docs/figure/var_type.Rmd/kron.compare-1.png" target="_blank">26d29a3</a>
</td>
<td>
Jason Willwerscheid
</td>
<td>
2019-08-16
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="results-distribution-of-p-values" class="section level2">
<h2>Results: Distribution of p-values</h2>
<p>A comparison of implied <span class="math inline">\(p\)</span>-value distributions (as described <a href="pvals.html">here</a>) offers additional evidence for the superiority of the gene-wise fit relative to the constant one. The fits that account for heteroskedasticity among both genes and cells (which I will refer to as “Kronecker-like” fits) show a bit further improvement.</p>
<pre class="r"><code>KL &lt;- sapply(lapply(res, `[[`, &quot;p.vals&quot;), `[[`, &quot;KL.divergence&quot;)
KL.df &lt;- data.frame(fit = names(KL), KL = KL)
ggplot(KL.df, aes(x = fit, y = KL)) +
  geom_bar(stat = &quot;identity&quot;) +
  scale_x_discrete(limits = the.limits) +
  labs(x = NULL, y = &quot;KL divergence relative to uniform&quot;)</code></pre>
<p><img src="figure/var_type.Rmd/pval.kl-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-pval.kl-1">
Past versions of pval.kl-1.png
</button>
</p>
<div id="fig-pval.kl-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/willwerscheid/scFLASH/blob/e2d99bb92ebb6b20e8700aefcb5d2bd837c2e290/docs/figure/var_type.Rmd/pval.kl-1.png" target="_blank">e2d99bb</a>
</td>
<td>
Jason Willwerscheid
</td>
<td>
2019-08-31
</td>
</tr>
<tr>
<td>
<a href="https://github.com/willwerscheid/scFLASH/blob/26d29a37b931efee69f83e3c05aa72a401c96948/docs/figure/var_type.Rmd/pval.kl-1.png" target="_blank">26d29a3</a>
</td>
<td>
Jason Willwerscheid
</td>
<td>
2019-08-16
</td>
</tr>
</tbody>
</table>
</div>
</div>
<div id="constant-variance-structure" class="section level3">
<h3>Constant variance structure</h3>
<pre class="r"><code>plot.p.vals(res$const$p.vals)</code></pre>
<p><img src="figure/var_type.Rmd/const.p-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-const.p-1">
Past versions of const.p-1.png
</button>
</p>
<div id="fig-const.p-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/willwerscheid/scFLASH/blob/e2d99bb92ebb6b20e8700aefcb5d2bd837c2e290/docs/figure/var_type.Rmd/const.p-1.png" target="_blank">e2d99bb</a>
</td>
<td>
Jason Willwerscheid
</td>
<td>
2019-08-31
</td>
</tr>
<tr>
<td>
<a href="https://github.com/willwerscheid/scFLASH/blob/26d29a37b931efee69f83e3c05aa72a401c96948/docs/figure/var_type.Rmd/const.p-1.png" target="_blank">26d29a3</a>
</td>
<td>
Jason Willwerscheid
</td>
<td>
2019-08-16
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="gene-wise-variance-structure" class="section level3">
<h3>Gene-wise variance structure</h3>
<pre class="r"><code>plot.p.vals(res$bygene$p.vals)</code></pre>
<p><img src="figure/var_type.Rmd/bygene.p-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-bygene.p-1">
Past versions of bygene.p-1.png
</button>
</p>
<div id="fig-bygene.p-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/willwerscheid/scFLASH/blob/e2d99bb92ebb6b20e8700aefcb5d2bd837c2e290/docs/figure/var_type.Rmd/bygene.p-1.png" target="_blank">e2d99bb</a>
</td>
<td>
Jason Willwerscheid
</td>
<td>
2019-08-31
</td>
</tr>
<tr>
<td>
<a href="https://github.com/willwerscheid/scFLASH/blob/26d29a37b931efee69f83e3c05aa72a401c96948/docs/figure/var_type.Rmd/bygene.p-1.png" target="_blank">26d29a3</a>
</td>
<td>
Jason Willwerscheid
</td>
<td>
2019-08-16
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="kronecker-variance-structure" class="section level3">
<h3>Kronecker variance structure</h3>
<p>I omit the Kronecker-like fits; to the eye, they are indistinguishable from the Kronecker fit.</p>
<pre class="r"><code>plot.p.vals(res$kron$p.vals)</code></pre>
<p><img src="figure/var_type.Rmd/kron.p-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-kron.p-1">
Past versions of kron.p-1.png
</button>
</p>
<div id="fig-kron.p-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/willwerscheid/scFLASH/blob/e2d99bb92ebb6b20e8700aefcb5d2bd837c2e290/docs/figure/var_type.Rmd/kron.p-1.png" target="_blank">e2d99bb</a>
</td>
<td>
Jason Willwerscheid
</td>
<td>
2019-08-31
</td>
</tr>
<tr>
<td>
<a href="https://github.com/willwerscheid/scFLASH/blob/26d29a37b931efee69f83e3c05aa72a401c96948/docs/figure/var_type.Rmd/kron.p-1.png" target="_blank">26d29a3</a>
</td>
<td>
Jason Willwerscheid
</td>
<td>
2019-08-16
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
</div>
<div id="results-log-likelihood-of-implied-distribution" class="section level2">
<h2>Results: Log likelihood of implied distribution</h2>
<p>Since the change-of-variables ELBO adjustments <a href="pseudocount.html#elbo_comparison">are not entirely reliable</a>, it’s useful to compare the log likelihoods of the <a href="pvals.html">implied discrete distributions</a> as well.</p>
<p>From this perspective, the Kronecker fit does best, followed by the Kronecker-like fit that pre-scales cells. I’m inclined to believe these statistics more than the ELBOs. In effect, since sparse genes have very small residual standard deviations, a <span class="math inline">\(p\)</span>-value prior on genes would imply that nonzero loadings are expected to be much larger for sparse genes than for genes that are highly expressed. A <span class="math inline">\(p\)</span>-value prior on cells doesn’t make much sense either, but the range of residual standard deviations for cells is much smaller, so there’s less of a difference between the <span class="math inline">\(p\)</span>-value prior and the usual prior.</p>
<pre class="r"><code>llik &lt;- sapply(lapply(res, `[[`, &quot;p.vals&quot;), `[[`, &quot;llik&quot;)
llik.df &lt;- data.frame(fit = names(llik), llik = llik)
ggplot(llik.df, aes(x = fit, y = llik)) + 
  geom_point() +
  scale_x_discrete(limits = c(&quot;const&quot;, &quot;bycell&quot;, &quot;bygene&quot;, &quot;kron&quot;, 
                              &quot;scale.genes&quot;, &quot;scale.cells&quot;, &quot;scale.both&quot;)) +
  labs(x = NULL, y = &quot;Log likelihood of implied distribution&quot;)</code></pre>
<p><img src="figure/var_type.Rmd/llik-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-llik-1">
Past versions of llik-1.png
</button>
</p>
<div id="fig-llik-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/willwerscheid/scFLASH/blob/9c8fdf9f3df59dd28461e450d639a009bdc77425/docs/figure/var_type.Rmd/llik-1.png" target="_blank">9c8fdf9</a>
</td>
<td>
Jason Willwerscheid
</td>
<td>
2019-09-03
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="results-factors-and-cell-types" class="section level2">
<h2>Results: Factors and cell types</h2>
<p>I plot the Kronecker factors in decreasing order of proportion of variance explained, and I attempt to arrange the factors from the other fits so that similar factors are aligned with one another. In particular, the plots visually confirm that the Kronecker-like factors are qualitatively very similar to the Kronecker factors (with the exception of factor 18).</p>
<p>At first, it seems like the gene-wise fit yields sparser factors than the Kronecker fit, but on closer inspection it looks to be producing a lot of factors that are driven by a few outlying ciliated cells, while the Kronecker fit yields more factors that identify substructure within the much larger population of basal and club cells.</p>
<p>The constant variance structure is obviously unsuitable.</p>
<pre class="r"><code>kron.order &lt;- order(res$kron$fl$pve, decreasing = TRUE)
plot.factors(res$kron,
             droplet$cell.type,
             kron.order,
             title = &quot;Kronecker variance structure&quot;)</code></pre>
<p><img src="figure/var_type.Rmd/plotting.factors-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-plotting.factors-1">
Past versions of plotting.factors-1.png
</button>
</p>
<div id="fig-plotting.factors-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/willwerscheid/scFLASH/blob/e2d99bb92ebb6b20e8700aefcb5d2bd837c2e290/docs/figure/var_type.Rmd/plotting.factors-1.png" target="_blank">e2d99bb</a>
</td>
<td>
Jason Willwerscheid
</td>
<td>
2019-08-31
</td>
</tr>
<tr>
<td>
<a href="https://github.com/willwerscheid/scFLASH/blob/26d29a37b931efee69f83e3c05aa72a401c96948/docs/figure/var_type.Rmd/plotting.factors-1.png" target="_blank">26d29a3</a>
</td>
<td>
Jason Willwerscheid
</td>
<td>
2019-08-16
</td>
</tr>
</tbody>
</table>
</div>
</div>
<pre class="r"><code>
kron.v.prescale &lt;- compare.factors(res$kron$fl, res$scale.cells$fl)
plot.factors(res$scale.cells,
             droplet$cell.type,
             kron.v.prescale$fl2.k[kron.order],
             title = &quot;Kronecker-like (pre-scale cells)&quot;)</code></pre>
<p><img src="figure/var_type.Rmd/plotting.factors-2.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-plotting.factors-2">
Past versions of plotting.factors-2.png
</button>
</p>
<div id="fig-plotting.factors-2" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/willwerscheid/scFLASH/blob/e2d99bb92ebb6b20e8700aefcb5d2bd837c2e290/docs/figure/var_type.Rmd/plotting.factors-2.png" target="_blank">e2d99bb</a>
</td>
<td>
Jason Willwerscheid
</td>
<td>
2019-08-31
</td>
</tr>
<tr>
<td>
<a href="https://github.com/willwerscheid/scFLASH/blob/26d29a37b931efee69f83e3c05aa72a401c96948/docs/figure/var_type.Rmd/plotting.factors-2.png" target="_blank">26d29a3</a>
</td>
<td>
Jason Willwerscheid
</td>
<td>
2019-08-16
</td>
</tr>
</tbody>
</table>
</div>
</div>
<pre class="r"><code>
kron.v.bygene &lt;- compare.factors(res$kron$fl, res$bygene$fl)
plot.factors(res$bygene, 
             droplet$cell.type, 
             kron.v.bygene$fl2.k[kron.order],
             title = &quot;Gene-wise variance structure&quot;)</code></pre>
<p><img src="figure/var_type.Rmd/plotting.factors-3.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-plotting.factors-3">
Past versions of plotting.factors-3.png
</button>
</p>
<div id="fig-plotting.factors-3" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/willwerscheid/scFLASH/blob/e2d99bb92ebb6b20e8700aefcb5d2bd837c2e290/docs/figure/var_type.Rmd/plotting.factors-3.png" target="_blank">e2d99bb</a>
</td>
<td>
Jason Willwerscheid
</td>
<td>
2019-08-31
</td>
</tr>
<tr>
<td>
<a href="https://github.com/willwerscheid/scFLASH/blob/26d29a37b931efee69f83e3c05aa72a401c96948/docs/figure/var_type.Rmd/plotting.factors-3.png" target="_blank">26d29a3</a>
</td>
<td>
Jason Willwerscheid
</td>
<td>
2019-08-16
</td>
</tr>
</tbody>
</table>
</div>
</div>
<pre class="r"><code>
kron.v.const &lt;- compare.factors(res$kron$fl, res$const$fl)
plot.factors(res$const, 
             droplet$cell.type,
             kron.v.const$fl2.k[kron.order],
             title = &quot;Constant variance structure&quot;)</code></pre>
<p><img src="figure/var_type.Rmd/plotting.factors-4.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-plotting.factors-4">
Past versions of plotting.factors-4.png
</button>
</p>
<div id="fig-plotting.factors-4" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/willwerscheid/scFLASH/blob/e2d99bb92ebb6b20e8700aefcb5d2bd837c2e290/docs/figure/var_type.Rmd/plotting.factors-4.png" target="_blank">e2d99bb</a>
</td>
<td>
Jason Willwerscheid
</td>
<td>
2019-08-31
</td>
</tr>
<tr>
<td>
<a href="https://github.com/willwerscheid/scFLASH/blob/26d29a37b931efee69f83e3c05aa72a401c96948/docs/figure/var_type.Rmd/plotting.factors-4.png" target="_blank">26d29a3</a>
</td>
<td>
Jason Willwerscheid
</td>
<td>
2019-08-16
</td>
</tr>
</tbody>
</table>
</div>
</div>
<br> <br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span> Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>#&gt; R version 3.5.3 (2019-03-11)
#&gt; Platform: x86_64-apple-darwin15.6.0 (64-bit)
#&gt; Running under: macOS Mojave 10.14.6
#&gt; 
#&gt; Matrix products: default
#&gt; BLAS: /Library/Frameworks/R.framework/Versions/3.5/Resources/lib/libRblas.0.dylib
#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/3.5/Resources/lib/libRlapack.dylib
#&gt; 
#&gt; locale:
#&gt; [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
#&gt; 
#&gt; attached base packages:
#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     
#&gt; 
#&gt; other attached packages:
#&gt; [1] flashier_0.1.15 ggplot2_3.2.0   Matrix_1.2-15  
#&gt; 
#&gt; loaded via a namespace (and not attached):
#&gt;  [1] Rcpp_1.0.1        plyr_1.8.4        highr_0.8        
#&gt;  [4] compiler_3.5.3    pillar_1.3.1      git2r_0.25.2     
#&gt;  [7] workflowr_1.2.0   iterators_1.0.10  tools_3.5.3      
#&gt; [10] digest_0.6.18     evaluate_0.13     tibble_2.1.1     
#&gt; [13] gtable_0.3.0      lattice_0.20-38   pkgconfig_2.0.2  
#&gt; [16] rlang_0.3.1       foreach_1.4.4     parallel_3.5.3   
#&gt; [19] yaml_2.2.0        ebnm_0.1-24       xfun_0.6         
#&gt; [22] withr_2.1.2       stringr_1.4.0     dplyr_0.8.0.1    
#&gt; [25] knitr_1.22        fs_1.2.7          rprojroot_1.3-2  
#&gt; [28] grid_3.5.3        tidyselect_0.2.5  glue_1.3.1       
#&gt; [31] R6_2.4.0          rmarkdown_1.12    mixsqp_0.1-119   
#&gt; [34] reshape2_1.4.3    ashr_2.2-38       purrr_0.3.2      
#&gt; [37] magrittr_1.5      whisker_0.3-2     MASS_7.3-51.1    
#&gt; [40] codetools_0.2-16  backports_1.1.3   scales_1.0.0     
#&gt; [43] htmltools_0.3.6   assertthat_0.2.1  colorspace_1.4-1 
#&gt; [46] labeling_0.3      stringi_1.4.3     pscl_1.5.2       
#&gt; [49] doParallel_1.0.14 lazyeval_0.2.2    munsell_0.5.0    
#&gt; [52] truncnorm_1.0-8   SQUAREM_2017.10-1 crayon_1.3.4</code></pre>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
http://docs.mathjax.org/en/latest/configuration.html.  This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
